includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+require_once( AMP__DIR__ . '/includes/sanitizers/class-amp-base-sanitizer.php' );
+
+/**
+ * Collects inline styles and outputs them in the amp-custom stylesheet.
+ */
+class AMP_Style_Sanitizer extends AMP_Base_Sanitizer {
+	private $styles = array();
+
+	public function sanitize() {
+		$body = $this->get_body_node();
+		$this->collect_styles_recursive( $body );
+
+		if ( ! empty( $this->styles ) ) {
+			add_action( 'amp_post_template_css', array( $this, 'append_styles' ), 99 );
+		}
+	}
+
+	private function collect_styles_recursive( $node ) {
+		if ( $node->nodeType !== XML_ELEMENT_NODE ) {
+			return;
+		}
+
+		if ( $node->hasAttributes() && $node instanceof DOMElement ) {
+			$style = $node->getAttribute( 'style' );
+			$class = $node->getAttribute( 'class' );
+
+			if ( $style ) {
+				$style = $this->process_style( $style );
+
+				if ( $style ) {
+					$class_name = $this->generate_class_name( $style );
+					$new_class  = trim( $class . ' ' . $class_name );
+
+					$node->setAttribute( 'class', $new_class );
+					$this->styles[ $class_name ] = $style;
```

Comment:
Another approach could be to _not_ remove the style attribute in Blacklist, and just handle everything in the Style class. It seems like this could be confusing to other devs, though, since the Blacklist is designed specifically for removing tags and attributes...